numpy/lib/format.py

```diff
@@ -915,3 +919,130 @@ def _read_bytes(fp, size, error_template="ran out of data"):
         raise ValueError(msg % (error_template, size, len(data)))
     else:
         return data
+
+class AppendArray:
+    def __init__(self, filename):
+        self.filename = filename
+        self.fp = None
+        self.__is_init = False
+        if os.path.isfile(filename):
+            self.__init()
+
+    def __create_header_bytes(self):
+        header_map = {
+            'descr': dtype_to_descr(self.dtype),
+            'fortran_order': self.fortran_order,
+            'shape': tuple(self.shape)
+        }
+        io = BytesIO()
+        write_array_header_2_0(io, header_map)
+
+        # create array header with 64 byte space space for shape to grow
+        io.getbuffer()[8:12] = pack("<I", int(io.getbuffer().nbytes-12+64))
+        io.getbuffer()[-1] = 32
+        io.write(b" "*64)
+        io.getbuffer()[-1] = 10
```

Comment:
I think you should add an argument to `write_array_header_2_0` to encapsulate this, either as:
* `min_header_size`, which sets the minimum header padding size
* `extra_header_padding`, which augments the header with extra padding.

Reply:
I think the extra_header_padding is what we need. The question is whether to make it an argument or just add 64 byte by default, since as mentioned, more energy would be needed to populate such an array than boiling earth's oceans. In a distant future, when our descendants will build a dyson swarm, this value may be increased to 128 byte to account for all particled in the known universe.
Adding 64 byte by default will increase the size of every .npy file by 64 bytes, but I cannot really judge how much of a problem that would be. Maybe none at all, maybe filesystem overhead adds 64 byte anyway, maybe this would pose some issues to certain users that have millions/billions of .npy files laying around. What do you think?

Reply:
> more energy would be needed to populate such an array than would be necessary to boil earth's oceans. In a distant future, when our descendants eventually build a dyson swarm, this value may be increased to 128 byte to account for all particles in the known universe.

I think you're thinking about this the wrong way. The length of an array is stored in an int64, so we're already limited to 2**63-1. The question is what the longest string is that is less than that size; for instance, `(9223372036854775807,)` is shorter than `(10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10)`. On the other hand, I think the way you have appending set up means that only one dimension of the array can grow, in which case `len(str(2**63-1)) - len(str(len(arr)))` is the needed amount of padding

Reply:
So perhaps `write_array_header_2_0(..., extra_padding=True)` and have `write_array_header_2_0` deal with calculating how much padding is needed.

Reply:
I have chosen to try the -1 option, see below. Then we would not need to handle possible growing headers in the first place.

Reply:
That's true, the more dimensions and larger dimensions the array has, the more bytes are covered by advancing along the first axis. So the worst case is an array with only one dimension of dtype int8 or so. So `len(str(2**63-1))` (=19) would be an upper bound. That means we could also just take 19 bytes as a buffer instead of 64 bytes. I would propose to make the code as simple as possible, so one could just add 19 bytes and not try to reduce that byte count by taking into account the influence of the other axes. For most arrays 19 spare bytes would probably not even make a difference in the file size, although even if, it would probably be consumed by the filesystem overhead.
I'll do it in a follow up pull request, first this pull request here (which does not affect anything else) should go through (as far as I understand it).

Reply:
Actually, shouldn't it be 2^64-1? I mean `size_t` is a `uint64_t` in C (for 64 bit machines). This would increase the buffer from 19 to 20 bytes.